gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java

```diff
@@ -153,12 +124,96 @@ static GoogleCloudStorageGrpcReadChannel open(
     }
   }
 
+  private static GoogleCloudStorageGrpcReadChannel openChannel(StorageStubProvider stubProvider,
+      StorageResourceId resourceId, GoogleCloudStorageReadOptions readOptions,
+      BackOffFactory backOffFactory) throws IOException {
+    StorageBlockingStub stub = stubProvider.newBlockingStub();
+    // TODO(b/135138893): We can avoid this call by adding metadata to a read request.
+    //      That will save about 40ms per read.
+    com.google.google.storage.v1.Object storageObject = getObjectMetadata(resourceId, readOptions,
+        stub);
+
+    Preconditions.checkArgument(storageObject != null, "object metadata cannot be null");
+    int prefetchSizeInBytes = readOptions.getMinRangeRequestSize() / 2;
+    long footerOffsetInBytes = Math.max(0, (storageObject.getSize() - prefetchSizeInBytes));
+
+    ByteString footerContent = getFooterContent(resourceId, readOptions, stub, footerOffsetInBytes);
+
+    return new GoogleCloudStorageGrpcReadChannel(
+        stub,
+        stubProvider,
+        resourceId,
+        storageObject.getGeneration(),
+        storageObject.getSize(),
+        footerOffsetInBytes,
+        footerContent,
+        readOptions,
+        backOffFactory);
+  }
+
+  private static com.google.google.storage.v1.Object getObjectMetadata(StorageResourceId resourceId,
+      GoogleCloudStorageReadOptions readOptions, StorageBlockingStub stub) throws IOException {
+    try {
+      // TODO(b/151184800): Implement per-message timeout, in addition to stream timeout.
+      GetObjectRequest getObjectRequest = GetObjectRequest.newBuilder()
+          .setBucket(resourceId.getBucketName())
+          .setObject(resourceId.getObjectName())
+          .build();
+      com.google.google.storage.v1.Object storageObject = stub
+          .withDeadlineAfter(readOptions.getGrpcReadMetadataTimeoutMillis(), MILLISECONDS)
+          .getObject(getObjectRequest);
+      // The non-gRPC read channel has special support for gzip. This channel doesn't
+      // decompress gzip-encoded objects on the fly, so best to fail fast rather than return
+      // gibberish unexpectedly.
+      if (storageObject.getContentEncoding().contains("gzip")) {
+        throw new IOException(
+            "Can't read GZIP encoded files - content encoding support is disabled.");
+      }
+      return storageObject;
+    } catch (StatusRuntimeException e) {
+      throw convertError(e, resourceId);
+    }
+  }
+
+  private static ByteString getFooterContent(StorageResourceId resourceId,
+      GoogleCloudStorageReadOptions readOptions, StorageBlockingStub stub, long footerOffset)
+      throws IOException {
+    try {
+      Iterator<GetObjectMediaResponse> footerContentResponse = stub
```

Comment:
We don't do much logging in this file. I think we should add logs at important decision area, such as (a) reading footer, size etc.., (b) footer cache hit (and miss) etc..

Reply:
Done